File: server/services/payment_service.py
from config.db_config import ADVANCE_PAYMENT_RATE
from utils.helpers import generate_demo_transaction_ref
from services.commission_service import calculate_commission
from models.models import Payment, Booking

def calculate_payment_breakdown(total_amount, advance_rate=None):
    """
    Computes advance requirement (20%) and remaining balance (80%).
    """
    rate = advance_rate if advance_rate is not None else ADVANCE_PAYMENT_RATE
    advance_amount = round(total_amount * rate, 2)
    remaining_amount = round(total_amount - advance_amount, 2)
    
    return {
        'total_amount': round(total_amount, 2),
        'advance_rate': rate,
        'advance_amount': advance_amount,
        'remaining_amount': remaining_amount
    }

def process_demo_payment(db, booking, payment_type='ADVANCE', payment_method='UPI (Simulated)'):
    """
    Executes simulated advance payment for hackathon demo.
    Generates EH-DEMO-XXXXXX reference, records payment in DB, and confirms booking.
    """
    if payment_type == 'ADVANCE':
        amount_to_pay = float(booking.advance_amount)
    elif payment_type == 'REMAINING':
        amount_to_pay = float(booking.remaining_amount)
    else:
        amount_to_pay = float(booking.total_amount)
        
    txn_ref = generate_demo_transaction_ref()
    
    payment = Payment(
        booking_id=booking.id,
        amount=amount_to_pay,
        payment_type=payment_type,
        payment_status='PAID',
        payment_method=payment_method,
        transaction_reference=txn_ref
    )
    db.add(payment)
    
    # Update Booking Status
    booking.booking_status = 'CONFIRMED'
    booking.payment_status = 'PAID' if payment_type in ['FULL', 'ADVANCE'] else 'PAID'
    
    # Calculate EventHub commission
    comm = calculate_commission(float(booking.total_amount), float(booking.commission_rate))
    booking.commission_amount = comm['commission_amount']
    
    db.commit()
    db.refresh(payment)
    db.refresh(booking)
    
    return {
        'payment': payment.to_dict(),
        'booking': booking.to_dict(),
        'transaction_reference': txn_ref,
        'message': 'Payment successful! Booking confirmed.'
    }

File: server/services/paymentService.js
/**
 * Dummy Advance Payment Service
 * Realistic simulated payment processing with transaction references.
 */

const { query } = require('../config/db');

const DEFAULT_ADVANCE_PERCENTAGE = 20.0; // 20% advance payment required for confirmation

const calculateAdvanceAmount = (totalAmount, percentage = DEFAULT_ADVANCE_PERCENTAGE) => {
  const total = Number(totalAmount) || 0;
  const advance = Math.round((total * percentage) / 100);
  const remaining = total - advance;

  return {
    totalAmount: total,
    advancePercentage: percentage,
    advanceAmount: advance,
    remainingAmount: remaining
  };
};

const generateTransactionRef = () => {
  const randomNum = Math.floor(100000 + Math.random() * 900000);
  return `EH-DEMO-${randomNum}`;
};

const processDemoPayment = async ({
  bookingId,
  amount,
  paymentMethod = 'UPI (Simulated)',
  paymentType = 'ADVANCE',
  notes = 'Simulated Advance Payment'
}) => {
  const transactionRef = generateTransactionRef();

  // 1. Insert Payment Record
  const insertSql = `
    INSERT INTO payments (booking_id, transaction_ref, amount, payment_type, payment_method, payment_status, notes)
    VALUES (?, ?, ?, ?, ?, 'PAID', ?)
  `;
  await query(insertSql, [bookingId, transactionRef, amount, paymentType, paymentMethod, notes]);

  // 2. Update Booking Status to CONFIRMED and payment_status to PAID
  const updateBookingSql = `
    UPDATE bookings 
    SET booking_status = 'CONFIRMED', payment_status = 'PAID'
    WHERE id = ?
  `;
  await query(updateBookingSql, [bookingId]);

  return {
    success: true,
    transactionRef,
    amount: Number(amount),
    paymentType,
    paymentMethod,
    paymentStatus: 'PAID',
    bookingStatus: 'CONFIRMED',
    timestamp: new Date().toISOString(),
    message: 'Demo payment processed successfully. Booking confirmed!'
  };
};

module.exports = {
  DEFAULT_ADVANCE_PERCENTAGE,
  calculateAdvanceAmount,
  generateTransactionRef,
  processDemoPayment
};
